### PR TITLE
WIP: Transform as plugin

### LINKED ISSUE #12674: Remove mapping source transform scripts
The ability to transform the `_source` field before indexing (while storing the original `_source`) is confusing and error prone.  Most of what you can do with it can already be done with `copy_to`.  It also makes it difficult to know where a field value has come from, which means it can't support highlighting.

I think this is not widely used at all and should be removed in 2.0 (while continuing to support indices which already use it for 2.x releases)

### PR DESCRIPTION
**Pull request by nik9000**
Moves transform script into a plugin.

Still very much a work in progress.

Closes #12674

### CODE REVIEW

**Comment by nik9000** [comment]
@jpountz I think the only way I can extract transform to a plugin without massive surgery is to support arbitrary transformations implementable by plugins. The mapping-transform plugin I've started on would simply by the "script" based implementation of that. And Elasticsearch would have no builtin implementation. Potentially `copy_to` is a builtin implementation if we wanted to spend lots of time on this, but I don't think we do.

**Comment by nik9000** [comment]
I'd love to get some opinions on if moving the ScriptTransform out of core is enough to call this "not supported by core". I'm just not sure of a clean way of moving the transform hooks out of core.

**Comment by rjernst** [comment]
My problem is the mapper service still needs the script service. I think we agree these transformations just shouldn't be done at this level. Could we either instead work on a way to add pre-processing to document injest (outside of mappings), or just drop this entirely and tell users they should use a real ETL before sending the document, or with something like logstash?

**Comment by nik9000** [comment]
Oh!  I think I can drop that direct dependency. It'd be mapper depends on a
thing from a plugin which in turn depends on scripeservice.

Users certainly should use a real etl pipeline in most cases. I still think
elasticsearch is the appropriate place to transform stuff when the result
doesn't belong in the source. It's rare that that is a good thing but
sometimes it is.

I'm happy to move it out of mapper into some other level but I don't know a
level to move it to.

File: core/src/main/java/org/elasticsearch/index/mapper/MapperServiceModule.java [hunk 1]
```diff
@@ -19,15 +19,48 @@
...
+
+import java.util.Set;
 
 /**
- *
+ * The infrastructure around DocumentMappers and FieldMappers.
  */
 public class MapperServiceModule extends AbstractModule {
+    // TODO it should be possible to move all of the rootTypeParsers into rootParsers as "extensions" registered in Elasticsearch core.
+    private static final Set<String> ROOT_TYPE_PARSERS = ImmutableSet.of(IndexFieldMapper.NAME, SourceFieldMapper.NAME,
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/index/mapper/MapperServiceModule.java]: Can we not use the old "root mapper" terminology? These are "metadata" fields now.

File: docs/plugins/mapping.asciidoc [hunk 2]
```diff
@@ -1,10 +1,14 @@
-[[mapper]]
-== Mapper Plugins
+[[mapping]]
```
Comment (rjernst) [inline docs/plugins/mapping.asciidoc]: Does this move need to be added to the "renames" doc (I don't remember exactly what it is called)?

File: core/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java [hunk 3]
```diff
@@ -128,18 +128,8 @@ public Builder put(MetadataFieldMapper.Builder<?, ?> mapper) {
...
-        }
-
-        /**
-         * @deprecated Use {@link #transform(ScriptService, Script)} instead.
-         */
-        @Deprecated
-        public Builder transform(ScriptService scriptService, String script, ScriptType scriptType, String language,
-                Map<String, Object> parameters) {
-            sourceTransforms.add(new ScriptTransform(scriptService, new Script(script, scriptType, language, parameters)));
+        public Builder transform(SourceTransform sourceTransform) {
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java]: This is the other problem I have. Transforms aren't really pulled out of mappings, the implementation is just moved out (which makes this *more* configurable and harder to remove).

**Comment by rjernst** [comment]
Moving script service out of mapper service would make this more bearable, but I still think we should just drop the ability to do this altogether. An ETL, before elasticsearch is the right place to do transformations. I don't see why you would want transformations to not be in _source. That would mean every time you reindex the document you need to reapply the transform. If there is a possibility you may want to change the transformed value (say you have a category field that you are mapping to an integer code), then store the original value as well, and you can still reorder codes when reindexing.

**Comment by rjernst** [comment]
And to be clear, my concrete suggestion here is we completely remove this feature, without a plugin. We can rethink the feature, or ability to do manipulation of document pre/post _source and decide what to do there in a later release. But I don't think we should keep around this feature in its current state for the sake of backcompat, it is too new a feature, that IMO is a failed experiment.

**Comment by nik9000** [comment]
> But I don't think we should keep around this feature in its current state for the sake of backcompat, it is too new a feature, that IMO is a failed experiment.

Fair enough. I'm tired of defending this. I don't know of anyone that is using it in a way that they can't just move to a proper transformation pipeline. It fills a niche that nothing else fills but no one needs that niche. And that niche is advanced enough that if you truly need it you can come on github and argue with us to get it back in. At least then we'd have a non-theoretical "good" use case.

**Pull request closed without merging**